### DISCUSSION TITLE
How to add explicit `mixed` type?

### DISCUSSION THREAD

**Discussion by brandonkelly** [rectorphp/rector#7033] [Q&A] [answered]
I need to create a rule that adds explicit `mixed` type declarations to properties and methods using `AddPropertyTypeDeclarationRector`, `AddReturnTypeDeclarationRector`, and `AddParamTypeDeclarationRector`, e.g.

```diff
- public $foo;
+ public mixed $foo;
```

```diff
- public function foo($bar)
+ public function foo(mixed $bar): mixed
```

I’ve tried passing `PHPStan\Type\MixedType` and `PHPStan\Type\StrictMixedType`, but neither of them are actually adding the type declaration to the code.

For example:

```php
<?php
declare(strict_types = 1);

use PHPStan\Type\MixedType;
use Rector\TypeDeclaration\Rector\ClassMethod\AddReturnTypeDeclarationRector;
use Rector\TypeDeclaration\ValueObject\AddReturnTypeDeclaration;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function(ContainerConfigurator $containerConfigurator): void {
    $services = $containerConfigurator->services();

    $services->set(AddReturnTypeDeclarationRector::class)
        ->configure([
            new AddReturnTypeDeclaration('my\Class', 'foo', new MixedType()),
        ]);
};
```

Is there a way to achieve this?

**Comment by samsonasik** [comment]
mixed type on property is currently not supported as far as I know due to can be various implementation:

- `@var mixed` is marked as mixed
- no type marked as mixed
- `private mixed $property` is marked as mixed as well.

that will require a very careful implementation, on very specific use case, and require check child/parent use case for check above use cases.

You can however, create your own rule for it to fill it:

```php
        $node->type = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode(new MixedType(true), TypeKind::PROPERTY());
        return $node;
```

**Comment by samsonasik** [comment, marked as answer]
@brandonkelly I stand corrected, it actually supported on `AddPropertyTypeDeclarationRector`, you can define explicit mixed type `new MixedType(true)` like the following:

```php
return static function (ContainerConfigurator $containerConfigurator): void {
    $services = $containerConfigurator->services();

    $services->set(AddPropertyTypeDeclarationRector::class)
        ->configure([
            new AddPropertyTypeDeclaration('my\Class', 'value', new MixedType(true)),
        ]);
};
```

**Reply by brandonkelly** [reply]
Doh… can’t believe I didn’t see the `$isExplicitMixed` arg 🤦🏻 

Confirmed it works with all three of those rules! Thanks a lot!
